cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/ReduceAggregator.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.api;
+
+import io.cdap.cdap.api.annotation.Beta;
+
+/**
+ * Groups all input objects into collections, and performs an aggregation on the entire group.
+ * Objects that have the same group key are placed into the same group for aggregation.
+ * It has better performance than {@link Aggregator} since the number of group values is reduced in each split before
+ * grouping all the values for the group key.
+ *
+ * For example, to aggregate and compute the average for the values, the plugin will first group all the values based
+ * on the group key, considering it generates following splits:
+ * Split 1: (key1, 1), (key1, 2), (key1, 3), (key2, 4)
+ * Split 2: (key1, 2), (key1, 3), (key1, 4), (key2, 4)
+ * Split 3: (key1, 3), (key1, 4), (key1, 5), (key2, 4)
+ * The reduce function will be called in each split to generate following:
+ * Split 1: (key1, sum: 6, count: 3), (key2, sum: 4, count: 1)
+ * Split 2: (key1, sum: 9, count: 3), (key2, sum: 4, count: 1)
+ * Split 3: (key1, sum: 12, count: 3), (key2, sum: 4, count: 1)
+ * Then reduce function is called when grouping all values in the split:
```

Comment:
Data is then shuffled across the cluster such that records with the same key are handled by the same executor:

Split 4: (key1, sum: 6, count: 3), (key1, sum:9, count:3), (key1, sum:12, count:3)
Split 5: (key2, sum:4, count:1), (key2, sum:4, count:1), (key2, sum:4, count:1)

The reduce function is called again to generate:

Split 4: (key1, sum:27, count:9)
Split 5: (key2, sum:12, count:3)

Finally, the transform method is called to generate the final output value(s):

Split 4: (key1, avg: 3)
Split 5: (key2, avg: 3)